enhancements/hypershift/konnectivity-socks5-proxy.md

```diff
@@ -0,0 +1,124 @@
+authors:
+  - "@awgreene"
+reviewers:
+  - "@kevinrizza"
+  - "@njhale"
+approvers:
+  - "@kevinrizza"
+creation-date: 2021-09-21
+last-updated: 2021-09-21
+status: implementable
+see-also:
+  - "/enhancements/update/ibm-public-cloud-support.md"
+replaces:
+  -
+superseded-by:
+  -
+---
+
+# Konnectivity Socks5 Proxy
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of this enhancement is to provide [HyperShift](https://github.com/openshift/hypershift/) control plane services the ability to establish [GRPC](https://grpc.io/) connections with servers running on guest clusters.
+
+## Motivation
+
+The HyperShift solution deploys OpenShift clusters by hosting their control plane as pods in a central Management Cluster. This type of setup allows the hosting of many cluster control planes on a single Kubernetes cluster. Each Control Plane has a related Guest Cluster where user workloads run.
+
+The [Operator Lifecycle Manager (OLM)](https://github.com/operator-framework/operator-lifecycle-manager) is a member of the control plane and runs within the Management Cluster. One requirement for OLM is to support Custom CatalogSources defined by users on the Guest Cluster.
+To meet this requirement, OLM needs to have the ability to establish GRPC connections with CatalogSource services running on the Guest Cluster.
+
+As of today, OLM is unable to connect to CatalogSources defined on the Guest Cluster because:
+- The Services defined on the guest clusters cannot be resolved by the DNS Service on the Management Cluster.
+- OLM cannot connect to CatalogSources running on the Guest Cluster.
+
+For OLM to support custom Catalog Sources it must be able to establish connections with CatalogSources running on the Guest Cluster.
+
+### Goals
+
+The primary goal of this enhancement is to propose a solution that allows services running on the control plane to:
+- Resolve domain names using the DNS service on the Guest Cluster
+- Tunnel requests to services running on the Guest Cluster
+
+### Non-Goals
+
+- N/A
+
+## Proposal
+
+This enhancement proposes the introduction of a [Socks5 proxy](https://en.wikipedia.org/wiki/SOCKS) that forwards requests from the control plane to the guest cluster.
+
+The Control Plane already includes a [Konnectivity Server](https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/) which can be used to tunnel requests to IP addresses running on the Guest Cluster. Rather than requiring each component to support establishing a connection with the Konnectivity Server, this enhancement proposes the introduction of a Socks5 proxy that:
```

Comment:
Couldn't we run the konnektivity server itself as sidecar?

Reply:
This is a great question and one that I had to reach out to @csrwng for clarification.

The worker nodes on the guest cluster need to connect to a single konnectivity server so it knows to route traffic to them. Multiple konnectivity servers would mean that each node would need to connect to each of them.